File: ts/instrumenter/instrument.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { instrument } from "./instrument.mjs";

const sample = `const price: number = 40;
let tax = 3
const total = applyTax(price, tax);
const [a, b] = [1, 2];
var sum, rest;
for (let i = 0; i < 3; i++) {
	console.log("iter", i);
}
for (const v of [1, 2]) {
	console.error("v", v);
}
console.log("total:", total);
function applyTax(price: number, tax: number): number {
	const inner = price + tax;
	return inner;
}
`;

test("probes simple declarations, top level included", () => {
	const result = instrument(sample, "file:///main.ts", true, [], 1);
	assert.ok(result.generated);
	assert.equal(
		(sample.match(/\n/g) ?? []).length,
		(result.generated.match(/\n/g) ?? []).length,
	);
	assert.match(result.generated, /price: number = 40;; __atomis_probe\(/);
	assert.match(result.generated, /"price", price\);/);
	// ASI case: no original semicolon, ours is added before the call
	assert.match(result.generated, /let tax = 3; __atomis_probe\(/);
	assert.match(result.generated, /const inner = price \+ tax;; __atomis_probe\(/);
	const supported = result.probes.filter((probe) => probe.supported);
	const unsupported = result.probes.filter((probe) => !probe.supported);
	// price, tax, total, inner (destructuring + multi-decl unsupported;
	// for-loop declarations are not VariableStatements and stay untouched)
	assert.equal(supported.length, 4);
	assert.equal(unsupported.length, 2);
	assert.ok(!/let i = 0; __atomis_probe/.test(result.generated));
});

test("log markers pick the console stream and track loops", () => {
	const result = instrument(sample, "file:///main.ts", true, [], 7);
	assert.match(
		result.generated,
		/__atomis_log_loop\(1, 7, 7, 2, 6, 1, "i", i\);/,
	);
	assert.match(
		result.generated,
		/__atomis_log_loop\(2, 7, 10, 2, 9, 1, "v", v\);/,
	);
	assert.match(result.generated, /__atomis_log\(1, 7, 12, 1\);/);
});

test("manual mode inserts only selected ids", () => {
	const all = instrument(sample, "file:///main.ts", true, [], 1);
	const price = all.probes.find((probe) => probe.name === "price");
	const none = instrument(sample, "file:///main.ts", false, [], 1);
	assert.ok(!none.generated.includes("__atomis_probe"));
	const selected = instrument(
		sample,
		"file:///main.ts",
		false,
		[price.probeId],
		1,
	);
	assert.match(selected.generated, /"price", price\);/);
});

test("parse errors carry positions and skip generation", () => {
	const result = instrument(
		"const x = ;\n",
		"file:///main.ts",
		true,
		[],
		1,
	);
	assert.equal(result.generated, undefined);
	assert.equal(result.parseDiagnostics[0].line, 1);
	assert.ok(result.parseDiagnostics[0].column >= 10);
});

test("plain JavaScript parses with the JS script kind", () => {
	const result = instrument(
		"const x = 5\nconsole.log(x)\n",
		"file:///main.js",
		true,
		[],
		1,
	);
	assert.match(result.generated, /const x = 5; __atomis_probe\(/);
});

test("instrumented sources pass through unchanged", () => {
	const source = 'const x = 1; __atomis_probe("a", 1, 7, "x", x);\n';
	const result = instrument(source, "file:///main.ts", true, [], 1);
	assert.equal(result.generated, source);
	assert.equal(result.probes.length, 0);
});

test("a concise arrow body log gains its marker without becoming a statement", () => {
	const source = "const xs = [1, 2];\nxs.forEach((x) => console.log(x));\n";
	const result = instrument(source, "file:///main.ts", true, [], 1);
	assert.match(
		result.generated,
		/\(x\) => \(console\.log\(x\), __atomis_log\(1, 1, 2, 19\)\)/,
	);
	// Still a valid program: reinstrumenting parses it (the instrumenter
	// reports parse errors instead of output when it does not).
	const again = instrument(result.generated, "file:///main.ts", true, [], 1);
	assert.equal(again.generated, result.generated);
});

test("tsx parses as tsx instead of erroring on JSX", () => {
	const source = "const el = <div a={1}>hola</div>;\nexport default el;\n";
	const result = instrument(source, "file:///main.tsx", true, [], 1);
	assert.equal(result.parseDiagnostics.length, 0);
	assert.match(result.generated ?? "", /__atomis_probe\(/);
});

File: go/instrumenter/instrument_test.go
package main

import (
	"go/parser"
	"go/token"
	"strings"
	"testing"
)

const sample = `package main

import "fmt"

func main() {
	price := 40
	var tax int = 3
	total := applyTax(price, tax)
	a, b := 1, 2
	for i := 0; i < 3; i++ {
		fmt.Println("iter", i)
	}
	for _, v := range []int{1, 2} {
		fmt.Println("v", v)
	}
	log := total
	_ = log
	_ = a
	_ = b
}

func applyTax(price int, tax int) int {
	return price + tax
}
`

func TestProbesSimpleDeclarations(t *testing.T) {
	result := Instrument(sample, "file:///main.go", true, nil, 1)
	if !result.HasGenerated {
		t.Fatalf("expected generated output: %+v", result.ParseDiagnostics)
	}
	if strings.Count(sample, "\n") != strings.Count(result.Generated, "\n") {
		t.Fatal("line count changed")
	}
	if !strings.Contains(result.Generated, `price := 40; __atomis_probe(`) {
		t.Fatalf("price probe missing:\n%s", result.Generated)
	}
	if !strings.Contains(result.Generated, `"price", price)`) {
		t.Fatal("probe arguments missing")
	}
	if !strings.Contains(result.Generated, `var tax int = 3; __atomis_probe(`) {
		t.Fatal("var probe missing")
	}
	supported := 0
	unsupported := 0
	for _, probe := range result.Probes {
		if probe.Supported {
			supported++
		} else {
			unsupported++
		}
	}
	// price, tax, total, log (a,b multi-assign is unsupported; loop i is an
	// init statement and must be skipped entirely)
	if supported != 4 {
		t.Fatalf("expected 4 supported probes, got %d", supported)
	}
	if unsupported != 1 {
		t.Fatalf("expected 1 unsupported probe, got %d", unsupported)
	}
	if strings.Contains(result.Generated, "i := 0; __atomis_probe") {
		t.Fatal("for-init must not be probed")
	}
}

func TestLogMarkersTrackLoops(t *testing.T) {
	result := Instrument(sample, "file:///main.go", true, nil, 7)
	if !strings.Contains(result.Generated, `__atomis_log_loop(1, 7, 11, 3, 10, 2, "i", i)`) {
		t.Fatalf("for-loop marker missing:\n%s", result.Generated)
	}
	if !strings.Contains(result.Generated, `"v", v)`) {
		t.Fatal("range value marker missing")
	}
}

func TestManualModeInsertsOnlySelected(t *testing.T) {
	all := Instrument(sample, "file:///main.go", true, nil, 1)
	var priceID string
	for _, probe := range all.Probes {
		if probe.Name == "price" {
			priceID = probe.ProbeID
		}
	}
	none := Instrument(sample, "file:///main.go", false, nil, 1)
	if strings.Contains(none.Generated, "__atomis_probe") {
		t.Fatal("no probes expected with auto-inspect off")
	}
	selected := Instrument(sample, "file:///main.go", false, []string{priceID}, 1)
	if !strings.Contains(selected.Generated, `"price", price)`) {
		t.Fatal("manual probe missing")
	}
}

func TestParseErrorsCarryPositions(t *testing.T) {
	result := Instrument("package main\n\nfunc main() {\n\tx :=\n}\n", "file:///main.go", true, nil, 1)
	if result.HasGenerated {
		t.Fatal("expected parse failure")
	}
	if len(result.ParseDiagnostics) == 0 || result.ParseDiagnostics[0].Line < 4 {
		t.Fatalf("expected diagnostics at line 4+: %+v", result.ParseDiagnostics)
	}
}

func TestInstrumentedSourcePassesThrough(t *testing.T) {
	source := "package main\n\nfunc main() { x := 1; __atomis_probe(\"a\", 1, 1, \"x\", x) }\n"
	result := Instrument(source, "file:///main.go", true, nil, 1)
	if result.Generated != source || len(result.Probes) != 0 {
		t.Fatal("expected passthrough")
	}
}

func TestTypeSwitchAndSelectHeadersStayUntouched(t *testing.T) {
	source := `package main

import "fmt"

func describe(x any) {
	switch v := x.(type) {
	case int:
		fmt.Println(v)
	}
}

func pick(ch chan int) {
	select {
	case v := <-ch:
		fmt.Println(v)
	default:
	}
}
`
	result := Instrument(source, "file:///main.go", true, nil, 1)
	if !result.HasGenerated {
		t.Fatalf("expected generated output: %+v", result.ParseDiagnostics)
	}
	if strings.Contains(result.Generated, "x.(type); __atomis_probe") {
		t.Fatalf("probe spliced into the type switch guard:\n%s", result.Generated)
	}
	if strings.Contains(result.Generated, "<-ch; __atomis_probe") {
		t.Fatalf("probe spliced into the select comm clause:\n%s", result.Generated)
	}
	mustParse(t, result.Generated)
}

// The assertion every instrumenter needs most: whatever was spliced in,
// the output is still a Go program.
func mustParse(t *testing.T, generated string) {
	t.Helper()
	fset := token.NewFileSet()
	if _, err := parser.ParseFile(fset, "generated.go", generated, 0); err != nil {
		t.Fatalf("generated output does not parse: %v\n%s", err, generated)
	}
}

func TestEveryGeneratedSampleParses(t *testing.T) {
	for _, source := range []string{sample} {
		result := Instrument(source, "file:///main.go", true, nil, 1)
		if !result.HasGenerated {
			t.Fatalf("expected generated output: %+v", result.ParseDiagnostics)
		}
		mustParse(t, result.Generated)
	}
}
